tephra-api/src/main/java/co/cask/tephra/Transaction.java

```diff
@@ -108,18 +157,60 @@ public long getFirstShortInProgress() {
     return firstShortInProgress;
   }
 
+  /**
+   * Returns true if the given version corresponds to a transaction that was in-progress at the time this transaction
+   * started.
+   */
   public boolean isInProgress(long version) {
     return Arrays.binarySearch(inProgress, version) >= 0;
   }
 
+  /**
+   * Returns true if the given version is present in one of the arrays of excluded versions (in-progress and
+   * invalid transactions).
+   */
   public boolean isExcluded(long version) {
     return Arrays.binarySearch(inProgress, version) >= 0
       || Arrays.binarySearch(invalids, version) >= 0;
   }
 
+  /**
+   * Returns true if the the given version corresponds to one of the checkpoint versions in the current
+   * transaction.
+   */
+  public boolean isCheckpoint(long version) {
+    return Arrays.binarySearch(checkpointWritePointers, version) >= 0;
+  }
+
+  /**
+   * Returns whether or not the given version should be visible to the current transaction.  A version will be visible
+   * if it was successfully committed prior to the current transaction starting, or was written by the current
+   * transaction (using either the current write pointer or the write pointer from a prior checkpoint).
+   *
+   * @param version the data version to check for visibility
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   *
+   * @see #isVisible(long, boolean) to exclude the current write pointer from visible versions.  This method always
+   *      includes the current write pointer.
+   */
   public boolean isVisible(long version) {
+    return isVisible(version, true);
+  }
+
+  /**
+   * Returns whether or not the given version should be visible to the current transaction.  A version will be visible
+   * if it was successfully committed prior to the current transaction starting, or was written by the current
+   * transaction (using either the current write pointer or the write pointer from a prior checkpoint).
+   *
+   * @param version the data version to check for visibility
+   * @param excludeCurrentWritePointer whether writes from the current write pointer should be visible
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   */
+  public boolean isVisible(long version, boolean excludeCurrentWritePointer) {
```

Comment:
I am glad we all agree that we are mimicking SQL semantics. I relational databases, the visibility constraints or isolation level is typically configured for the life time of a transaction. I don't know of a way way in SQL to change the isolation level midway into a transaction (and if there is one, that is not a common pattern). So why would we want to set that for each operation? 

One advantage of making it a property of  the transaction is that semantics of a transaction are clean and straight-forward.  The other advantage is that we do not need to add a flag to every operation of every participating transaction-aware dataset, in order to participate in a check-pointed transaction.

Actually, coming back to @JamesRTaylor's remark on the Kira, in Phoenix, statements to not see their own writes. The way I understood that is that Phoenix will checkpoint the tx after every statement. My understanding was also that this applies to all statements in a transaction, including the first statement. @ghelmling I am not sure why you think that a transaction should start with "read-own-writes" and switch to "exclude-own-writes" at the time of the first Select-Upsert statement. If the first statement is a Select-Upsert, it should also not see its own writes, or should it?